Add exclude-paths option to dependabot.yml

dnlfm opened the pull request:
This PR adds the possibility of specifying `exclude-paths` in **dependabot.yml**.

It's a very old issue, and I decided to give a first stab: https://github.com/dependabot/dependabot-core/issues/4364

End-to-end tests may be required to validate the solution. Unit tests are in place.

Example of dependabot.yml with the new option:
```yaml
version: 2
updates:
  - package-ecosystem: "npm_and_yarn"
    directory: "/"
    schedule:
      interval: "daily"
    exclude-paths:
      - "src/test/assets"
      - "vendor/**"
      - "src/*.rb"
      - "src/test/helper.rb"
```

dnlfm commented:
> @dnlfm You may like to fix the specs which are failing.

@randhircs I'm sorry but why did you force push, rewriting previous commits? 
The failing spec is not related to the changes in this PR: 
<img width="1318" height="410" alt="image" src="https://github.com/user-attachments/assets/5286e417-48ed-4c1b-b29f-1771d086e37a" />

randhircs commented:
> > @dnlfm You may like to fix the specs which are failing.
> 
> @randhircs I'm sorry but why did you force push, rewriting previous commits?
> 
> The failing spec is not related to the changes in this PR: <img alt="image" width="1318" height="410" src="https://private-user-images.githubusercontent.com/36580837/469442580-5286e417-48ed-4c1b-b29f-1771d086e37a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMzOTA4NTIsIm5iZiI6MTc1MzM5MDU1MiwicGF0aCI6Ii8zNjU4MDgzNy80Njk0NDI1ODAtNTI4NmU0MTctNDhlZC00YzFiLWIyOWYtMTc3MWQwODZlMzdhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI0VDIwNTU1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlMTZmYmI0NjA4ZDZmMTU5NzRhYWI2MGNlNDRjMzNmODA5OWJkZjczZmExZTcxNjE3YWExMzE5NWFmYmRkNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iqT-QntOzhziOYCDcToYOrXiIKnFQBhky5dNwkrbluA">
> 
> It seems this issue appeared after your commit. Force-pushes should be avoided, it may have modified something that it is hard to know what it is, as it may have rewritten the git tree. The `Specs / ci` was fine before this happened: https://github.com/dependabot/dependabot-core/actions/runs/16314785084/job/46077982547

@dnlfm I had to rebase to proceed to check if all specs are passing. I am sure sorbet fix is something you can check from your end.

randhircs commented:
@dnlfm Again now, I am checking by rebasing it to see if all specs are good.

dnlfm commented:
> > > @dnlfm Again now, I am checking by rebasing it to see if all specs are good. i feel that you may like to fix the sorbet issue.
> > 
> > 
> > Hi @randhircs, I see "Sorbet" was added to this project but not really added to the files, i.e. not completely rolled out it seems. For example,
> > ```
> > Checking updater/lib/dependabot/file_fetcher_command.rb for Sorbet typing...
> > No Sorbet typing found at the top of updater/lib/dependabot/file_fetcher_command.rb
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The file `file_fetcher_command.rb` wasn't added by me, and it contains this error for Sorbet typing. I would like you to check that, as it is something that was not added in this PR. Another PR that I saw where Sorbet is failing but didn't add the file is: #12589 where the file `uv/lib/dependabot/uv/file_fetcher.rb` originally didn't have the Sorbet typing, confirming that it is something in the project that wasn't rolled out well.
> > I would like to also ask you to remove the need of Sorbet typing in test files, as I added `# typed: ignore`, it is still requiring to use `strict` or `strong`, but again, it doesn't make sense to have it in test files.
> > I would prefer to not have to check these Sorbet issues by myself.
> 
> @dnlfm — We actually added full Sorbet type checks later on. However, files without proper Sorbet typings won’t be caught by the pipeline unless they’re modified. Once someone touches such a file, it becomes a requirement to fully type it, with the goal of gradually achieving complete Sorbet coverage.
> 
> If you're able to add the Sorbet typings here, that would be great. But if it feels like too much, @randhircs can help with that. Thanks again for your contribution!
> 
> @randhircs — You might be able to help here. Also, if you're planning to push additional commits, please avoid force-pushing, as we’d like to retain @dnlfm’s original commit.

Hi @kbukum1,

Thank you for your response.

> it becomes a requirement to fully type it, with the goal of gradually achieving complete Sorbet coverage
I suspected that could be the case, but I would encourage a different approach: I would suggest that the team responsible for adding Sorbet could map the files that they need to go through and address each one of them, let's say, 40 files a week, prioritizing the ones in PRs that are already open. This way, PRs will have isolated changes, i.e. the PRs won't be touching something that is not meaningful to their changes, and thus making the code review process better.

Also, I would really appreciate if a code review is done on what I have done. I looked at my PR once again today, and I started wondering if what I have done will work for all kind of repositories or if there is something missing, for instance:
- I added the exclusion logic to `_fetch_repo_contents` and `_cloned_repo_contents`, covering the flow for `repo_contents` in the Base implementation of a FileFetcher. However, I noticed that the FileFetcher implementations also implements their own `fetch_files`. 
  I'm not very familiar with the whole flow, but I assumed that they would use `repo_contents` at some point and I believe it may cover what we want to achieve with these changes. But maybe someone with more knowledge of the codebase or that is able to make a E2E test can tell me if there's any issue or something else missing, and then we can continue from there.

randhircs commented:
@dnlfm We have provided fix for sorbet. You may like to resolve the conflict.

AbhishekBhaskar commented on common/lib/dependabot/config/file.rb:
```diff
@@ -34,14 +34,15 @@ def initialize(updates:, registries: nil)
       def update_config(package_manager, directory: nil, target_branch: nil)
         dir = directory || "/"
-        package_ecosystem = PACKAGE_MANAGER_LOOKUP.invert.fetch(package_manager)
+        package_ecosystem = REVERSE_PACKAGE_MANAGER_LOOKUP.fetch(package_manager, "dummy")
```
Can you also please explain what this change is for?

AbhishekBhaskar replied:
I'm reverting this change as well with regards to the above comment. Thanks.

dnlfm commented:
> @dnlfm, Also, we are planning on releasing this as a new feature. With regards to this, we would need to add a feature flag which when enabled, these changes would be in effect. I will be creating the feature flag shortly. Please let me know if you prefer adding the feature flag, which I can share with you, in the code or I can push those changes myself to the same PR. Thanks!

Hi @AbhishekBhaskar, that's great and thank you for asking! Please feel free to push the changes directly to this PR.

dnlfm commented on bin/docker-dev-shell:
```diff
@@ -281,6 +281,8 @@ docker run --rm -ti \
   -v "$(pwd)/updater/Gemfile:$CODE_DIR/dependabot-updater/Gemfile" \
   -v "$(pwd)/updater/lib:$CODE_DIR/dependabot-updater/lib" \
   -v "$(pwd)/updater/spec:$CODE_DIR/dependabot-updater/spec" \
+  -v "$(pwd)/Gemfile:$CODE_DIR/Gemfile" \
```
or was also to install dependencies in the container, e.g. `sorbet`

dnlfm replied:
@AbhishekBhaskar this was essential at least for me to do things inside the Docker container, that's why I've added here. It is not a problem to remove from this PR and add to a separate one, but I believe it is unnecessary.

mtrackeros commented:
> This PR adds the possibility of specifying `exclude-paths` in **dependabot.yml**.
> 
> It's a very old issue, and I decided to give a first stab: https://github.com/dependabot/dependabot-core/issues/4364
> 
> End-to-end tests may be required to validate the solution. Unit tests are in place.
> 
> Example of dependabot.yml with the new option:
> ```yaml
> version: 2
> updates:
>   - package-ecosystem: "npm_and_yarn"
>     directory: "/"
>     schedule:
>       interval: "daily"
>     exclude-paths:
>       - "src/test/assets"
>       - "vendor/**"
>       - "src/*.rb"
>       - "src/test/helper.rb"
> ```

dnlfm commented on updater/lib/dependabot/file_fetcher_command.rb:
```diff
@@ -121,9 +110,22 @@ def create_file_fetcher(directory: nil)
+      fetcher_args = args.merge(update_config: update_config)
+
+      Dependabot::FileFetchers.for_package_manager(job.package_manager).new(**fetcher_args)
     end
```
I see this PR is already merged, and I noticed this part was removed from the final version... 
I'm not very familiar with the whole codebase, but as I checked the flow for having the new flag, this piece of code was kinda "necessary".

I may be wrong of course, and since it wasn't mentioned, I'm wondering if any E2E test has been added to validate the changes in this PR or if they'll be added in a later moment. 
@AbhishekBhaskar @markhallen

dnlfm commented on common/lib/dependabot/file_fetchers/base.rb:
```diff
@@ -541,6 +554,66 @@ def _cloned_repo_contents(relative_path)
+
+        return entries if @exclude_paths.empty?
+
+        filtered_entries = entries.reject do |entry|
+          full_entry_path = entry.path
+          Dependabot.logger.info("DEBUG: Checking entry path: #{full_entry_path}")
+
+          @exclude_paths.any? do |exclude_pattern|
+            Dependabot.logger.info("DEBUG: Testing pattern: #{exclude_pattern} against path: #{full_entry_path}")
+
+            # case 1: exact match
+            exclude_exact = full_entry_path == exclude_pattern
+
+            # case 2: Directory prefix matching: check if path is inside an excluded directory
+            exclude_deeper = full_entry_path.start_with?("#{exclude_pattern}#{File::SEPARATOR}",
+                                                         "#{exclude_pattern}/")
+
+            # case 3: Explicit recursive (patterns that end with /**)
+            exclude_recursive = false
+            if exclude_pattern.end_with?("/**")
+              base_pattern = exclude_pattern[0...-3]
+              exclude_recursive = full_entry_path == base_pattern ||
+                                  full_entry_path.start_with?("#{base_pattern}/") ||
+                                  full_entry_path.start_with?("#{base_pattern}#{File::SEPARATOR}")
+            end
+
+            # case 4: Glob pattern matching with enhanced flags
+            # Use multiple fnmatch attempts with different flag combinations
+            fnmatch_flags = [
+              File::FNM_EXTGLOB,
+              File::FNM_EXTGLOB | File::FNM_PATHNAME,
+              File::FNM_EXTGLOB | File::FNM_PATHNAME | File::FNM_DOTMATCH,
+              File::FNM_PATHNAME
+            ]
+            exclude_fnmatch_paths = fnmatch_flags.any? do |flag|
+              File.fnmatch?(exclude_pattern, full_entry_path, flag)
+            end
+
+            result = exclude_exact || exclude_deeper || exclude_recursive || exclude_fnmatch_paths
+            Dependabot.logger.info("DEBUG: Pattern #{exclude_pattern} vs #{full_entry_path} -> #{result ? 'EXCLUDED' : 'INCLUDED'}") # rubocop:disable Layout/LineLength
```
I'm also wondering what was the reason behind removing the simple filtering code below:
```ruby
# Filters out any entries whose paths match one of the exclude_paths globs.
      sig { params(entries: T::Array[T.untyped]).returns(T::Array[T.untyped]) }
      def filter_excluded(entries)
        entries.reject do |entry|
          full_entry_path = entry.path
          @exclude_paths.any? do |ex|
            File.fnmatch?(ex, full_entry_path, File::FNM_EXTGLOB) || full_entry_path.start_with?(ex)
            exclude_exact   = full_entry_path == ex
            exclude_deeper  = full_entry_path.start_with?("#{ex}#{File::SEPARATOR}")
            File.fnmatch?(ex, full_entry_path, File::FNM_EXTGLOB) || exclude_exact || exclude_deeper
          end
        end
      end
```
I just want to know if there was any issue with that, that this new version covers?
@AbhishekBhaskar

dnlfm replied:
I'm asking mainly because no new tests were added to the unit test file regarding this filtering logic

dnlfm commented on common/spec/dependabot/file_fetchers/base_exclude_spec.rb:
```diff
@@ -0,0 +1,211 @@
+require "dependabot/config/update_config"
+require "dependabot/experiments"
+
+RSpec.describe Dependabot::FileFetchers::Base do
```
I've left a command at the top here (line 8) that has been removed:
```ruby
# Command to run this test: rspec common/lib/dependabot/file_fetchers/base_exclude_spec.rb
```
I don't see a reason to delete such command - it could help new contributors to test individual files

dnlfm replied:
Same for the changes I've made for adding volumes to docker-dev stuff - it is essential to have gemfile and Sorbet config in there, so we can do lint checks, etc...
